add _SwiftToolchain availability and add it to the SwiftStdlib macros

QuietMisdreavus opened the pull request:
rdar://151390924

The `swift` availability domain is used to determine the *language mode* that is required to use a particular symbol. However, it is still useful to know what toolchain release introduced a particular API, especially for documentation purposes. This is especially important for the `--{build,preview}-stdlib-docs` feature added in #76152, which currently only displays the potentially-misleading `swift` availability domain.

Many symbols in the standard library use the `SwiftStdlib` availability macros to indicate which versions of Apple's OSs they can be back-deployed to. While these macros are versioned according to the toolchain release version, this information is not conveyed into the AST at all.

This PR adds a `_SwiftToolchain` availability domain that can be used to indicate which version of the Swift toolchain a symbol was released in. It is meant to be added to availability macros like the `SwiftStdlib` macros, so it can be added to an `@available` attribute alongside other platforms, unlike `swift`. Its active version is set to the release version of the running compiler. The primary purpose for this domain is for documentation, so it's not allowed to be used in `#available` checks by itself. (It can be combined with platform domains in these situations, since availability macros are likely to be used in those situations as well.)

This PR then adds the `_SwiftToolchain` domain to the `SwiftStdlib` availability macros, and ensures that it can be properly rendered in symbol graphs when building the stdlib docs with `--{build,preview}-stdlib-docs`. (Right now the `SwiftToolchain` token is printed as-is in documentation; a separate PR to swift-docc-symbolkit is needed to render it more nicely.)

QuietMisdreavus committed:
add _SwiftToolchain availability and add it to the SwiftStdlib macros

rdar://151390924

QuietMisdreavus commented:
@swift-ci Please test

tshortli commented on lib/AST/AvailabilityDomain.cpp:
```diff
@@ -139,6 +145,7 @@ bool AvailabilityDomain::isActive(const ASTContext &ctx) const {
   switch (getKind()) {
   case Kind::Universal:
   case Kind::SwiftLanguage:
+  case Kind::SwiftToolchain:
```
I think this should be false so that we don't emit any diagnostics about `_SwiftToolchain` availability.

tshortli commented on lib/AST/AvailabilityConstraint.cpp:
```diff
@@ -124,7 +124,7 @@ static bool canIgnoreConstraintInUnavailableContexts(
     // conformances, which can sometimes be awkward to avoid references to.
     if (!isa<TypeDecl>(decl) && !isa<ExtensionDecl>(decl)) {
-      if (domain.isUniversal() || domain.isSwiftLanguage())
+      if (domain.isUniversal() || domain.isSwiftLanguage() || domain.isSwiftToolchain())
```
Since I don't think availability in the `_SwiftToolchain` domain should be diagnosed (in this release, at least) I don't think this should be updated.

tshortli commented:
It's a bit unfortunate that we're introducing a new hard-coded availability domain so close to having custom availability domains (https://forums.swift.org/t/pitch-extensible-availability-checking/79308). However, I don't think that will be ready in time to satisfy this use case so I agree that this is probably a reasonable thing to do anyways. I would like to have a path to removing the hard-coded domain in the future when custom domains are fleshed out, though.

tshortli commented:
I think this also needs to potentially require an experimental feature to be enabled in order to allow `@available(_SwiftToolchain ...)` to be written in source code. Otherwise, developers might add `@available(_SwiftToolchain)` to their own code, which I don't think we should commit to supporting yet.

QuietMisdreavus committed:
review: make _SwiftToolchain availability inactive

QuietMisdreavus committed:
review: remove SwiftToolchain check in unconditionally unavailable contexts

QuietMisdreavus committed:
strip toolchain availability from SwiftStdlib for tests

QuietMisdreavus committed:
SwiftifyMacro: use getNameForAttributePrinting and skip SwiftToolchain

QuietMisdreavus committed:
update tests to accommodate _SwiftToolchain additions

hnrklssn reviewed (approved):
SwiftifyImport part looks good

QuietMisdreavus committed:
drop SwiftToolchain availability from derived actor conformances

QuietMisdreavus commented:
@swift-ci Please test

QuietMisdreavus commented:
I rebased (a couple times) to deal with merge conflicts, and changed a couple related bits of collateral damage:

- I changed the SwiftifyMacro call (again) to specifically strip out `_SwiftToolchain` availability. It still calls `getNameForAttributePrinting`, though.
- I changed the implementation of `AvailabilityInference::applyInferredAvailableAttrs` (which is used to infer availability in a few places based on other decls) to optionally strip out `_SwiftToolchain` availability. This is then used in the derived conformances for `actor` and `distributed actor` decls.

Both of these changes were motivated by wanting to lessen the impact on module interface files outside the standard library (and also in tests) by not copying in `_SwiftToolchain` availability where it's not explicitly written.

@tshortli This all seemed unrelated to the work i talked with you about for gating this behind an experimental feature. I could still add that if you'd like.

QuietMisdreavus committed:
require experimental feature to use _SwiftToolchain

QuietMisdreavus commented:
@tshortli I've added the experimental feature requirement for the new availability domain. This should be the last of your review comments.

QuietMisdreavus committed:
add feature requirement to tests

QuietMisdreavus commented:
@swift-ci Please test

QuietMisdreavus committed:
fixup! add _SwiftToolchain availability and add it to the SwiftStdlib macros

QuietMisdreavus committed:
review: never infer toolchain availability

QuietMisdreavus commented:
@swift-ci Please test

QuietMisdreavus committed:
add SwiftToolchainAvailability to core runtime features

QuietMisdreavus commented:
@swift-ci Please test

etcwilde reviewed (changes requested):
While we are shipping the SDK as part of the toolchain bundle as a side-effect of the current state of the build system, toolchain versions don't determine what symbols are available from a runtime. Apple platforms build the runtimes separately from the toolchain and we are working toward splitting this for the other platforms as well.
I do think we need something to indicate the stdlib version that is separate from the OS version, but the current name is misleading.

QuietMisdreavus commented:
@etcwilde Are you suggesting that the current approach is entirely wrong, or do you just have a problem with the name? I will admit that i'm using `@available` for a different approach than its language semantics here, but there is currently no other source for this information that can be used for documentation without going to each API in the stdlib and adding either a new attribute or a doc comment with the DocC `@Available` directive, which is a massive-scale change.
